#**Introduction**
- Spam detection is an important application of machine learning and natural language processing (NLP) used to classify messages as Spam or Not Spam (Ham).

- Spam messages are unwanted or malicious messages sent in bulk for advertising, fraud, or phishing purposes.

###**Dataset Description**

The dataset contains text messages labeled as:

1. Spam (1): Unwanted promotional or fraudulent messages

2. Ham (0): Legitimate non-spam messages

In [22]:
# import libraries - for data processing and machine learning.
import numpy as np
import pandas as pd

In [23]:
# load dataset - for analysis and latin-1 encoding to handle special characters
df = pd.read_csv("/content/drive/MyDrive/ML-CSV/spam.csv", encoding='latin-1')

In [24]:
df.head()

,v1,v2,Unnamed: 2,Unnamed: 3,Unnamed: 4
0,ham,"Go until jurong point, crazy.. Available only ...",NaN,NaN,NaN
1,ham,Ok lar... Joking wif u oni...,NaN,NaN,NaN
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...,NaN,NaN,NaN
3,ham,U dun say so early hor... U c already then say...,NaN,NaN,NaN
4,ham,"Nah I don't think he goes to usf, he lives aro...",NaN,NaN,NaN


###**Data Preprocessing**
Cleans and prepares raw text data for accurate model training and prediction.

In [25]:
# dropping null values - Remove missing values and unnecessary columns.
df = df.drop(["Unnamed: 2", "Unnamed: 3", "Unnamed: 4"], axis = 1)

In [26]:
df.isnull().sum()

,0
v1,0
v2,0


In [27]:
# columns rename
df = df.rename(columns = {"v1":"target", "v2":"text"})

In [28]:
df.head()

,target,text
0,ham,"Go until jurong point, crazy.. Available only ..."
1,ham,Ok lar... Joking wif u oni...
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...
3,ham,U dun say so early hor... U c already then say...
4,ham,"Nah I don't think he goes to usf, he lives aro..."


In [29]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   target  5572 non-null   object
 1   text    5572 non-null   object
dtypes: object(2)
memory usage: 87.2+ KB


In [30]:
# label encoding - Convert spam/ham text labels into numeric values.
df = df.replace({"ham":"0", "spam": "1"})

In [31]:
df["target"].value_counts()

,count
target,
0,4825
1,747


In [32]:
df["text"].value_counts().head()

,count
text,
"Sorry, I'll call later",30
I cant pick the phone right now. Pls send a message,12
Ok...,10
"Wen ur lovable bcums angry wid u, dnt take it seriously.. Coz being angry is d most childish n true way of showing deep affection, care n luv!.. kettoda manda... Have nice day da.",4
Okie,4


In [33]:
# Input and Output
X = df['text']
y = df['target']

###**Feature Extraction**

Vectorization Techniques:-
1. CountVectorizer - Counts words
2. TF-IDF(Term Frequency-Inverse Document Frequency) - Gives importance to important words

In [34]:
# # Convert text into numerical data using CountVectorizer
# from sklearn.feature_extraction.text import CountVectorizer

# cv = CountVectorizer()
# X_vector = cv.fit_transform(X)

In [35]:
# Apply TF-IDF - Convert text messages into numerical feature vectors.
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer()
X_vector = tfidf.fit_transform(X)

##**Model Selection**

Choose suitable for model based on problem.

Like: 0, 1 -> Logistic Regression

In [36]:
# Train Test Split - Divide data into training and testing sets.
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X_vector, y, test_size=0.2, random_state=42)

In [37]:
# GaussianNB works with dense array
X_train_dense = X_train.toarray()
X_test_dense = X_test.toarray()

**Gaussian Naive Bayes**

A probabilistic classifier based on Bayes theorem.

Advantages:

1. Fast training
2. Works well for text classification
3. Handles large datasets efficiently

In [38]:
# Model - Train Gaussian Naive Bayes model using training data.
from sklearn.naive_bayes import GaussianNB

model_nb = GaussianNB()
model_nb.fit(X_train_dense, y_train)

GaussianNB()

In [39]:
# Prediction - Predict whether messages are spam or not spam.
y_pred = model_nb.predict(X_test_dense)

##**Model Evaluation**

Check performance based are:-

1. **Accuracy Score:** Measures how many predictions are correct out of all predictions.

Formula:

Accuracy= TP+TN+FP+FN / TP+TN

2. **Precision:** Measures how many predicted spam messages are actually spam.

Formula:

Precision = TP+FP / TP
	​

3. **Recall:** Measures how many actual spam messages are correctly detected.

Formula:

Recall = TP+FN / TP

4. **F1 Score:** Measures the balance between Precision and Recall.

Formula:

F1 Score = 2× Precision+Recall / Precision×Recall


In [46]:
# Check model accuracy and classification performance
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

print('Accuracy:', accuracy_score(y_test, y_pred))

conf_matrix = confusion_matrix(y_test, y_pred)
print("\nConfusion Matrix:\n",conf_matrix)

print('\nClassification Report:\n')
print(classification_report(y_test, y_pred))

Accuracy: 0.8968609865470852

Confusion Matrix:
 [[865 100]
 [ 15 135]]

Classification Report:

              precision    recall  f1-score   support

           0       0.98      0.90      0.94       965
           1       0.57      0.90      0.70       150

    accuracy                           0.90      1115
   macro avg       0.78      0.90      0.82      1115
weighted avg       0.93      0.90      0.91      1115



##**New Data - Predection**

In [41]:
# Predict new messages and conclude whether they are Spam or Not Spam
new_messages = [
    "Congratulations! You won ₹50,000. Click now to claim reward.",
    "Project meeting tomorrow at 10 AM. Please bring report."
]

predictions = model_nb.predict(tfidf.transform(new_messages).toarray())

for i in range(len(new_messages)):
    if predictions[i] == "1":
        print(new_messages[i], "--> Spam")
    else:
        print(new_messages[i], "--> Not Spam")

Congratulations! You won ₹50,000. Click now to claim reward. --> Spam
Project meeting tomorrow at 10 AM. Please bring report. --> Not Spam


##**Conclusion**

1. The spam detection model achieved **89.69% accuracy**, showing good overall prediction performance.

2. However, the model achieved **90% recall for spam**, meaning it successfully detected most spam messages.

3. The overall **91% weighted F1-score** confirms that the model is reliable and performs well for practical spam message classification.

4. The model correctly classified new messages as **Spam and Not Spam**, showing good prediction performance.